tasks/rbd_fio.py

```diff
@@ -0,0 +1,202 @@
+        fio_config.write('bs={bs}\n'.format(bs=bs))
+    else:
+        fio_config.write('bs=4k\n')
+    fio_config.write('iodepth=2\n')
+    if config.get('fio-io-size'):
+        size=config['fio-io-size']
+        fio_config.write('size={size}\n'.format(size=size))
+    else:
+        fio_config.write('size=100m\n')
+    
+    fio_config.write('time_based\n')
+    if config.get('runtime'):
+        runtime=config['runtime']
+        fio_config.write('runtime={runtime}\n'.format(runtime=runtime))
+    else:
+        fio_config.write('runtime=1800\n')
+    fio_config.write('allow_file_create=0\n')    
+    image_size=10240    
+    if config.get('image_size'):
+        image_size=config['image_size']
+        
+    formats=[1,2]
+    features=[1,2,4]
+    fio_version='2.2.9'
+    if config.get('formats'):
+        formats=config['formats']
+    if config.get('features'):
+        features=config['features']
+    if config.get('fio-version'):
+        fio_version=config['fio-version']
+    
+    fio_config.write('norandommap\n')
+    if ioengine == 'rbd':
+        fio_config.write('invalidate=0\n')
+    #handle package required for librbd engine
+    sn=remote.shortname
+    system_type= teuthology.get_system_type(remote)
+    if system_type == 'rpm' and ioengine == 'rbd':
+        log.info("Installing librbd1 devel package on {sn}".format(sn=sn))
+        remote.run(args=['sudo', 'yum' , 'install', 'librbd1-devel', '-y'])
+    elif ioengine == 'rbd':
+        log.info("Installing librbd devel package on {sn}".format(sn=sn))
+        remote.run(args=['sudo', 'apt-get', '-y', 'install', 'librbd-dev'])
+    if ioengine == 'rbd':
+        fio_config.write('clientname=admin\n')
+        fio_config.write('pool=rbd\n')
+    for frmt in formats:
+        for feature in features:
+           log.info("Creating rbd images on {sn}".format(sn=sn))
+           remote.run(args=['sudo', 'rbd', 'create',
+                            '--image', 'i{i}f{f}{sn}'.format(i=frmt,f=feature,sn=sn),
+                            '--image-format', '{f}'.format(f=frmt),
+                            '--image-features', '{f}'.format(f=feature),
+                            '--size', '{size}'.format(size=image_size)])
+           remote.run(args=['sudo', 'rbd', 'info', 'i{i}f{f}{sn}'.format(i=frmt,f=feature,sn=sn)])
+           if ioengine != 'rbd':
+               out=StringIO.StringIO()
+               remote.run(args=['sudo', 'rbd', 'map', 'i{i}f{f}{sn}'.format(i=frmt,f=feature,sn=sn)],stdout=out)
+               dev=re.search(r'(/dev/rbd\d+)',out.getvalue())
+               rbd_dev=dev.group(1)
+               if config.get('test-clone-io'):
+                    log.info("Testing clones using fio")
+                    remote.run(args=['sudo', 'rbd', 'snap', 'create', 'i{i}f{f}{sn}@i{i}f{f}{sn}Snap'.format(i=frmt,f=feature,sn=sn)])
+                    remote.run(args=['sudo', 'rbd', 'snap', 'protect', 'i{i}f{f}{sn}@i{i}f{f}{sn}Snap'.format(i=frmt,f=feature,sn=sn)])
```

Comment:
I notice repeating the same string templates many times. Wouldn't it be nicer to only do that once?